RFC-0018-torch-monitor.md

```diff
@@ -0,0 +1,253 @@
+    # high level
+    message: Optional[str]
+    # Timestamp is the unix wall clock time of the event in seconds
+    timestamp: Optional[float]
+    # Metadata can contain any event specific fields and is intended
+    # for fields to later be aggregated or filtered on.
+    metadata: Dict[str, Union[str, float, int, bool]]
+```
+
+All events are sent to all registered event handlers. Event handlers can ignore events that are not relevant to them. The backend for these events will be implemented in C++ and merged with the existing profiler where possible.
+
+Since handlers can be registered in Python, the frequency of logged events must be relatively low to avoid performance impact.
+
+```py
+event_handlers = []
+
+class EventHandler(ABC):
+    @abstractmethod
+    def handle(event: Event) -> None:
+        ...
+def register(handler: EventHandler):
+    global event_handlers
+    event_handlers.append(handler)
+
+def log(event: Event) -> None:
+    for handler in event_handlers:
+        handler.handle(event)
+```
+
+**Example Usage**
+
+```py
+from torch.monitor import Event, log
+
+@dataclass
+class RdzvEvent(Event):
+    def __init__(self, run_id: str, rank: int, message: str) -> None:
+        super().__init__(
+            type="elastic_rdvz",
+            metadata={"run_id": run_id, "rank": pid},
+            message=message,
+        )
+
+event = RdzvEvent(run_id="1234", rank=0, message="rendezvous started")
+log(event)
+```
+
+```py
+class ExampleJSONLHandler(EventHandler):
+    def handle(event: Event) -> None:
+        with open(self.file_name, 'a') as f:
+          json.dump(f, event.asdict())
+
+register(ExampleJSONLHandler('logs.jsonl'))
+```
+
+### Stats
+
+These stats are designed to be always on metrics to track various key metrics that can be used for monitoring and debugging the performance of your training jobs or inference systems.
+
+These are defined in Python for readability purposes. The core would be implemented in C++ for minimal overhead tracking of metrics and the ability to log more fine grained metrics automatically from things such as autograd or the optimizers.
+
+**Interfaces**
+
+```py
+class StatType(Enum):
+    # VALUE exports the most recently set value.
+    VALUE = "value"
+    # MEAN computes the mean of the set values within the window.
+    MEAN = "mean"
+    # COUNT tracks the number of times a value is set within the window.
+    COUNT = "count"
+    # SUM computes the sum of the values set within the window.
+    SUM = "sum"
+
+    MAX = "max"
+    MIN = "min"
+
+    # These may not be present in the initial implementation:
+
+    # HISTOGRAM computes summary statistics such as P50, P90, P95, P99
+    HISTOGRAM = "histogram"
+    # STDDEV computes the standard deviation of the values set within
+    # the window.
+    STDDEV = "stddev"
+
+_collectors: Set[StatCollector] = set()
+
+@dataclass
+class Stat:
+    # key is the name of the stat.
```

Comment:
Isn't counter a better name for this?

Shouldn't it include help and unit fields as well to make them more discoverable?